src/backend/common/dispatch.hpp

```diff
@@ -42,3 +44,73 @@ inline T greatestPrimeFactor(T n) {
 
     return v;
 }
+
+// For OPENCL, the dimensions of local are returned
+// usage: cl::NDRange local = bestBlockSize<cl::NDRange>(dims, WG)
+// For CUDA, the dimensions of 1 block are returned
+// usage: dim3 block = bestBlockSize<dim3>(dims, 32);
+// The parameter dims can have any type as long as it is convertable to unsigned
+
+// Remark: The bestBlockSize is only best for independent element operations, as
+// are: copying, scaling, math on independent elements, ...
+// Since vector dimensions can be returned, it is NOT USABLE FOR BLOCK
+// OPERATIONS, as are: matmul, etc.
+template<typename Tout, typename Tin>
+Tout bestBlockSize(const Tin dims[4], unsigned warp) {
+    const unsigned d0         = static_cast<unsigned>(dims[0]);
+    const unsigned d1         = static_cast<unsigned>(dims[1]);
+    const unsigned d2         = static_cast<unsigned>(dims[2]);
+    const unsigned OCC        = 3;
+    const unsigned elements   = d0 * d1;
+    const unsigned minThreads = warp / 4;  // quarter wave
+    const unsigned maxThreads =
+        std::min(warp * 4, divup(elements * warp, 16384U) * minThreads);
+
+    const unsigned threads0 =
+#ifdef AF_OPENCL
+        (d0 < warp) ? d0 :
+#endif
+        (d1 == 1) ? warp * 4
```

Comment:
I will, although I thought that the compiler optimization would do that for me.

Reply:
Updated to const unsigned warp4.

In a previous version, I had warp as a template variable so that it was optimized.  I changed this, with this as a consequence.